spec/models/project/stage_spec.rb

```diff
@@ -33,13 +33,112 @@
+      expect(subject.end_date).to eq(Date.parse("2024-11-27"))
+    end
+
+    it "sets end_date to start_date if a single date is provided" do
+      subject.date_range = "2024-11-26"
+      expect(subject.start_date).to eq(Date.parse("2024-11-26"))
+      expect(subject.end_date).to eq(Date.parse("2024-11-26"))
+    end
+  end
+
+  describe "#validate_date_range" do
+    it "is valid when both dates are blank" do
+      stage = build(:project_stage, start_date: nil, end_date: nil)
+      expect(stage).to be_valid
+    end
+
+    it "adds error if start_date is blank" do
+      subject.end_date = Time.zone.today
+      expect(subject).not_to be_valid
+      expect(subject.errors.symbols_for(:date_range)).to include(:incomplete)
+    end
+
+    it "adds error if end_date is blank" do
+      subject.start_date = Time.zone.today
+      expect(subject).not_to be_valid
+      expect(subject.errors.symbols_for(:date_range)).to include(:incomplete)
+    end
+
+    it "adds error if start_date is after end_date" do
+      subject.start_date = Date.tomorrow
+      subject.end_date = Time.zone.today
+      expect(subject).not_to be_valid
+      expect(subject.errors.symbols_for(:date_range)).to include(:start_date_must_be_before_end_date)
+    end
+
+    it "does not add errors if start_date is before or equal to end_date" do
+      subject.start_date = Time.zone.today
+      subject.end_date = Time.zone.today
+      expect(subject).not_to be_valid
+      expect(subject.errors[:date_range]).to be_empty
+    end
+
+    it "is invalid if type and class name do not match" do
+      subject.type = "Project::Gate"
+      expect(subject).not_to be_valid
+      expect(subject.errors.symbols_for(:type)).to include(:type_and_class_name_mismatch)
+    end
+  end
+
+  describe "#working_days_count" do
+    it "returns nil if not_set? is true" do
+      allow(Day).to receive(:working)
+
+      subject.start_date = nil
+      subject.end_date = nil
+
+      expect(subject.working_days_count).to be_nil
+      expect(Day).not_to have_received(:working)
+    end
+
+    it "returns the correct number of days if start_date and end_date are the same" do
+      subject.start_date = Time.zone.today
+      subject.end_date = Time.zone.today
+      expect(subject.working_days_count).to eq(1)
+    end
+
+    it "returns the correct number of days for a valid date range" do
+      subject.start_date = Date.parse("2024-11-25")
+      subject.end_date = Date.parse("2024-11-27")
+      expect(subject.working_days_count).to eq(3)
+    end
+
+    it "calls the Day.working.from_range method with the right arguments" do
+      subject.start_date = Date.parse("2024-11-25")
+      subject.end_date = Date.parse("2024-11-27")
+
+      allow(Day).to receive(:working).and_return(Day)
+      allow(Day).to receive(:from_range)
+                .with(from: subject.start_date, to: subject.end_date)
+                .and_return([])
+
+      expect(subject.working_days_count).to eq(0)
+    end
```

Comment:
Can you explain why you added this spec?

This spec is a bit odd as it tests the internals of the method in question. I wonder if it wouldn't be better to just supply some more test cases with specific date ranges and expected amounts of working days?

Right now, this spec would break if the `working_days_count` method changes its implementation, even if the result returned was the same.

Reply:
I see, that makes sense.

With that in mind, we should check if the allowed methods were actually called:
```ruby
      expect(Day).to have_received(:working).with(no_args)
      expect(Day).to have_received(:from_range).with(from: subject.start_date, to: subject.end_date)
```

Reply:
Very good point, for some reason that was left off 😅